consensus/view_change.go

```diff
@@ -98,30 +90,18 @@ func (pm *State) GetViewChangeDuraion() time.Duration {
 	return time.Duration(diff * diff * int64(viewChangeDuration))
 }
 
-// switchPhase will switch FBFTPhase to nextPhase if the desirePhase equals the nextPhase
-func (consensus *Consensus) switchPhase(desired FBFTPhase, override bool) {
-	if override {
-		consensus.phase = desired
-		return
-	}
-
-	var nextPhase FBFTPhase
-	switch consensus.phase {
-	case FBFTAnnounce:
-		nextPhase = FBFTPrepare
-	case FBFTPrepare:
-		nextPhase = FBFTCommit
-	case FBFTCommit:
-		nextPhase = FBFTAnnounce
-	}
-	if nextPhase == desired {
-		consensus.phase = nextPhase
-	}
-}
-
 // GetNextLeaderKey uniquely determine who is the leader for given viewID
-func (consensus *Consensus) GetNextLeaderKey() *bls.PublicKeyWrapper {
-	wasFound, next := consensus.Decider.NextAfter(consensus.LeaderPubKey)
+func (consensus *Consensus) GetNextLeaderKey(viewID uint64) *bls.PublicKeyWrapper {
+	gap := 1
+	if viewID > consensus.GetCurBlockViewID() {
+		gap = int(viewID - consensus.GetCurBlockViewID())
+	}
+	consensus.getLogger().Info().
+		Str("leaderPubKey", consensus.LeaderPubKey.Bytes.Hex()).
+		Uint64("newViewID", viewID).
+		Uint64("myCurBlockViewID", consensus.GetCurBlockViewID()).
+		Msg("[GetNextLeaderKey] got leaderPubKey from coinbase")
+	wasFound, next := consensus.Decider.NextAfter(consensus.LeaderPubKey, gap)
```

Comment:
consensus.LeaderPubKey is updated for every view change. This logic change will mess up the leader change. Let's just keep the same logic as a pure refactor?